### ISSUE TITLE
Create automatic tests of Dasharo build documentation

### ISSUE THREAD

**Issue by philipanda** [Dasharo/dasharo-issues#1153]
### The problem you're addressing (if any)

Every historic release of Dasharo has to be possible to be built by anyone with access to the sources. The resulting firmware must be identical to the ones we publish. The process of building Dasharo for all supported devices is described at docs.dasharo.com, but it turns out the instructions are not always completely flawless.
https://github.com/Dasharo/dasharo-issues/issues/1139
https://github.com/Dasharo/dasharo-issues/issues/1143

The CI at https://github.com/Dasharo/coreboot/ makes sure the firmware can be build on all the devices, but it only checks it on the `dasharo` branch and does not do it according to the documentation, like a person would, but according to a CI script, which is not guaranteed to do the same steps as described in the documentation.

### Describe the solution you'd like

An automatic test that will check if following the instructions from the documentation allows to build any historic release of Dasharo, for any device, which will be identical to the ones we publish.

The test should find all the published historic releases for every supported device, find the build documentations for the devices, build the firmware according to the instructions in the documentation and compare if the resulting binary is identical as the published release.

The build process should start from a fresh installation of a selected OS to make sure no dependency is missing from the documentation. Ideally from the https://github.com/dasharo/preseeds

### Where is the value to a user, and who might that user be?

The ability to reproduce the historic releases, which is not always guaranteed due to faults in documentation

### Additional context

Discussion and a simple implementation of such test by copying the build instructions manually: https://github.com/Dasharo/open-source-firmware-validation/pull/545

- Manual testing would be easy, but equally unreliable and time consuming as the reason for the issues with build documentation is human error. Multiple people have been creating and updating the build instructions over time. The instructions don't follow a strict pattern.  
- The test must be performed by reading the documentation in some way. Copying the documentation to a script does not make sense as both of them would need to be kept up to date at the same time. And there still would be a lot of possible human errors involved.
- The documentation might need to be adjusted to be able to be analyzed by a bot. The `.md` or `.html` files could be analyzed to find out the build instructions. Either:
    - All the codeblocks should contain only valid commands, mindlessly following which would result in building the FW
    - Some well defined syntax should be used to identify the blocks to execute and/or places where things like the release revision has to be inserted.
- Using `Jinja or another template language` was suggested as a way of creating the build documentation to minimize the human error when creating it. The same templates could be used as an input to the test. 
- Simplifying the build process with a universal build script was suggested as a way to decrease the human error and define the build process well. Such script would need to support all historical releases too. A trial at extending the `build.sh` script in coreboot repo to fulfill this task: https://github.com/Dasharo/coreboot/pull/579. Using this script is not ideal as it is a part of the coreboot repository, which needs to be checked out to release revisions, which won't contain the script. If a build tool like that would be simple enough, so that the build process caould be performed in one command, the tests of the documentation, and the documentation as a whole could be simplified a lot. 
- The build can be tested on qemu, although adding it to the CI of the documentation repository should be done carefully as building Dasharo so many times could easily take too long. One build in https://github.com/Dasharo/coreboot CI easily takes over 5 minutes.

**Comment by macpijan** [comment]
The discussion here is still ongoing on many fronts. We can narrow down the scope if there will be interest in approaching this problem. Please start the discussion here first prior picking this up.

**Comment by krystian-hebel** [comment]
Perhaps https://github.com/antmicro/tuttest could be useful here?

**Comment by macpijan** [comment]
Another possibly relevant project: https://doc-detective.com/

**Comment by macpijan** [comment]
> Perhaps https://github.com/antmicro/tuttest could be useful here?

Similar project - maybe a bit more active, as the first one is 2 yrs with no commits in main branch: https://github.com/earldouglas/codedown

**Comment by shinymonitor** [comment]
Instead of using a dsl or external dependency, we can use a json schema.  For example, something like:
[
  {
    "step": "Install Dependencies",
    "variants": [
      {
        "constraints": {"builder_platform": "debian", "architecture": ["i386", "aarch64"]},
        "commands": [
          "sudo apt-get update",
          "sudo apt-get install -y gcc-aarch64-linux-gnu"
        ]
      },
      {
        "constraints": {},
        "_comment": "Default fallback",
        "commands": ["echo Please install dependencies manually"]
      }
    ]
  },
  {
    "step": "Build",
    "variants": [
      {
        "constraints": {"vendor": "msi"},
        "commands": ["make build-{model}"]
      },
      {
        "constraints": {},
        "commands": ["make default"]
      }
    ]
  }
]

then we also pass a target attributes for example something like:
{"builder_platform": "arch", "device_platform": "debian", "architecture": "aarch64", "vendor": "msi", "model": "XYZ"}

this will match to default fallback for dependency install step and the first one for the build step. it will also do string substitution for attributes (for example we can do "make crossgcc-{architecture} CPUS=$(nproc)" for the toolchain building step)

generating docs from this json is easy. to generate the json from docs automatically we need to write another tool (like codedown). from next release we can add this schema into docs and their generation.

to test binaries from previous releases, we can make a meta program that generates target attributes and generated binaries and match them with released ones.

i have a working runner and a simple docs generator for a pr.

**Comment by Clubhouse1661** [comment]
Hi, I'm interested in working on this bounty.                                                                        

I see you've explored several approaches.
Could you clarify if there's a preferred direction at this point, or would you like me to evaluate the options and propose a solution?

Also, if this is available, I'd appreciate being assigned to the issue.

**Comment by philipanda** [comment]
Hi @Clubhouse1661, thanks for joining in.

The issue looks like an open discussion and I don't think there was a preferred option decided somewhere out of record. Do you know of any @macpijan?

I think the decision of what tool will be chosen is up to the contributor as long as it satisfies the issue. There's definitely an unwritten rule to stick to FOSS solutions as that's compatible with the company's vision. And obviously we need to respect the licenses.
It would definitely be a great feat if you could describe the thought process, by for example highlighting the pros/cons of different tools.

**Comment by macpijan** [comment]
>  Do you know of any @macpijan?

There is no specified solution. As long as it does the job - tests commands from documentation - it is welcome to be proposed here.

**Comment by philipanda** [comment]
Are there any progress or difficulties you'd like to share @Clubhouse1661?

**Comment by philipanda** [comment]
@Clubhouse1661 You'll be unassigned from this issue due to lack of activity so that others get a chance with it. If you're still interested in the bounty you can ask to be assigned again just like before

**Comment by WuXieSec** [comment]
I prepared a phase-1 implementation and submitted PR:
https://github.com/Dasharo/docs/pull/1240

What it delivers now:
- automated CI check for build-doc command snippets (non-executing)
- syntax validation of shell/bash blocks in `docs/variants/**/building-manual.md`
- placeholder/pseudocode-aware handling to avoid false negatives

This is intended as an incremental baseline for #1153 and can be extended with controlled execution in a follow-up phase.

**Comment by copernicusjones** [comment]
I'd like to work on this bounty. My approach:

1. Parse building-manual.md pages from the Dasharo docs repo to extract build commands
2. Build a Python test framework that:
   - Spins up a Docker container matching Ubuntu 22.04 (as specified in preseeds)
   - Clones the correct Dasharo coreboot repo at the release tag from the docs
   - Executes the build commands from the documentation inside the container
   - Compares the resulting .rom binary SHA256 against the published release hash
3. Run this as a GitHub Actions CI job in the docs repo that triggers on PRs and weekly cron

Starting the implementation now. Will update with progress.

**Comment by ameyypawar** [comment]
@macpijan — you asked to discuss here before picking this up, so before going further I'd like to check the direction. I put together a small proof-of-concept as something concrete to react to (Dasharo/open-source-firmware-validation#1276, draft), and it surfaced a few things worth deciding.

**On "parse the docs" vs "standardise the build".** The building manuals are MkDocs Material *tabbed decision trees* (firmware × device × options via `content.tabs.link`), so the "run every code block" approach tools like doc-detective/codedown take concatenates mutually-exclusive branches — likely part of why they misbehave. Resolving a single path per selection works on the current docs with no changes, and as a side effect can flag where a manual is ambiguous (version choices buried in prose, checkout/build tabs that don't line up). That said, it doesn't compete with the universal-`build.sh` idea (Dasharo/coreboot#579) — if the build collapses to one command, a resolver just targets that instead. The two directions seem complementary rather than either/or.

**On the harder half.** The doc-detective run you tried failed on a missing host dependency (`sudo`) — which is exactly the "fresh OS, nothing missing" part of this issue, and the valuable, unsolved bit. The PoC deliberately doesn't attempt it (no fresh-OS provisioning, no build execution, no release/hash discovery yet); it only covers reading the docs into a concrete recipe.

**Reconciling with existing work.** @WuXieSec's Dasharo/docs#1240 already checks build-docs commands in the docs repo. I'd rather converge with that than duplicate it. Where would you want this to live — osfv, or docs alongside Dasharo/docs#1240 — and is this direction worth continuing? Happy to reshape or drop it based on your call.

**Comment by FloatingPegasus** [comment]
@macpijan Would a separately scoped US$75 milestone for QEMU Q35 release-artifact discovery be useful alongside the existing documentation-test PRs?

The concrete gap is resolving the correct published input: v0.1.0 is in `Dasharo/edk2` with DEBUG/RELEASE CODE/VARS assets and checksum sidecars, while v0.2.x is in `Dasharo/coreboot` with normal/all-menus ROM variants. I propose an explicit variant selector plus fixture-based tests for repository/tag resolution, ambiguous or missing assets, and available checksum metadata. This would supply inputs to the existing verification work; the integration point would be agreed with you and its authors first.

Before implementation, could you confirm whether this separate milestone is wanted, funded, and assignable, and whether disclosed Codex-assisted implementation and review are acceptable? I would also need to confirm an available PayPal payout route privately. This is a scope inquiry, not a claim on the full issue or a contribution submission.

**Comment by mowafymxis** [comment]
Hi, I'm checking eligibility before considering this documentation-testing bounty. I'm 17 and based in Egypt. Does the program accept contributors my age, and would guardian consent be required? I would use disclosed AI assistance for implementation and testing; please let me know whether that is acceptable.

Could an approved reward be sent through your Open Collective/Wise process directly to an Egyptian bank account in my own name, arriving in EGP, without me opening a PayPal or Wise account?

I see the existing documentation-test PRs and proposals, so I don't want to duplicate their work. Is there any remaining funded scope suitable for a new contributor, and what reward would apply? I haven't claimed the issue or started implementation. Thank you.

**Comment by stantheman0128** [comment]
Following the request to discuss scope here first, would a separately assigned **US$100 gross** pilot for **fresh-OS build-documentation validation** be useful?

I read the current QEMU Q35 building manual and OSFV #1276. The latter explicitly leaves fresh-OS dependency capture and actual builds out of scope; the existing release-artifact proposal addresses a different input. I would focus on one agreed QEMU Q35 release and the normal `qemu` variant, using a disposable Ubuntu 22.04 VM rather than a container that silently supplies host dependencies.

Proposed acceptance:
- Start from an agreed clean OS image and record the base package inventory, documentation revision, firmware tag and build-container identity.
- Follow the manual and its explicitly linked Docker setup prerequisites, applying only the agreed `REVISION` substitution. Record missing dependencies or ambiguous prose instead of silently repairing the environment.
- Complete one real firmware build, retain the command/exit-status log and output hash, and provide a repeatable run procedure plus any narrowly necessary documentation correction. An actual clean-OS blocker would be reported as a blocker, not a passing build.
- Agree the input interface with the existing documentation-extraction work so this does not duplicate its parser or the separate release-discovery proposal. This pilot would establish the execution stage before expanding to every historic release.

Would you fund and assign this milestone, and should its harness live in OSFV or the docs repository? Please confirm the completion criterion if the first run exposes a substantive historical build failure rather than a documentation omission. I understand your published support ranges are discretionary and the amount above is only a proposal, not an existing award.

The work would use disclosed OpenAI Codex assistance. I have only read public source/docs so far and would wait for scope/assignment before main implementation. Please confirm that this workflow is eligible. I understand payout uses Open Collective; formal program-term acceptance and any expense request would be handled by the account owner, not by this inquiry, and no private payment details need to be posted here.

**Comment by SproutSeeds** [comment]
Is there a funded build-documentation milestone on #1153 that is still available separately from the existing pilot proposals?

I could validate one QEMU Q35 release on a clean supported Linux image, document missing dependencies or failing build steps, and compare the output with the published firmware. Before starting, I'd confirm the release, OS, reproducibility expectations, and acceptance checks.

I use AI tools for investigation and automated tests, and handle review and required manual validation myself. Would that workflow be acceptable, and what reward and payment terms would apply? If that scope overlaps existing work, I'd be interested in a remaining piece.

Thanks,
Cody
